File: analysis/06_nhl/nhl_loop.py
"""The NHL prop analysis loop: is the projection any good, and (once the ledger fills) does
betting it against posted lines make money?

  A. Paper ROI by market x side x strength, flat $1, bootstrap 95% CI. Empty until the
     season opens (2026-10-07); prints and moves on.
  B. Slices a rule can act on: edge band, market, side (Wilson CI on hit, flat ROI).
  C. Walk-forward projection calibration on the stored game logs (same recipe as
     nhl_edge.calibrate: shrinkage toward league average, recent-10 tilt, Poisson): log-loss
     vs the naive league-average model and a reliability table for shots / points / saves.
     This runs today on 46k player-games and is the pre-season evidence.

Output: analysis/_out/nhl_roi.csv, nhl_slices.csv, nhl_calibration.csv.
Mirrors nhl_loop.R — keep in lockstep. Point estimates must match; bootstrap CIs may differ
in the last digit.
"""
from __future__ import annotations

import math
import sys
from pathlib import Path

import numpy as np
import pandas as pd

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from _shared.load_data import wilson
from _shared.load_nhl import load_game_logs, load_nhl_bets

OUT_DIR = Path(__file__).resolve().parents[1] / "_out"
MIN_BETS = 10
MIN_N = 5
BOOT_REPS = 5000
SEED = 20261007

# mirror nhl_edge.py — a change there must be reflected here
PRIOR_SEASON = 20252026
SHRINK_GAMES = 20.0
RECENT_GAMES = 10
RECENT_WEIGHT = 0.35
MIN_PRIOR = 10
LINES = {"shots": 2.5, "points": 0.5, "saves": 27.5}

EDGE_BINS, EDGE_LABELS = [0, 8, 15, 30, 50, 100000], ["0-8", "8-15", "15-30", "30-50", "50+"]


def verdict(lo: float, hi: float) -> str:
    return "PROFITABLE (95% CI > 0)" if lo > 0 else "losing (95% CI < 0)" if hi < 0 else "inconclusive"


def boot_ci(x: np.ndarray, rng: np.random.Generator) -> tuple[float, float]:
    idx = rng.integers(0, len(x), size=(BOOT_REPS, len(x)))
    means = x[idx].mean(axis=1)
    return float(np.percentile(means, 2.5)), float(np.percentile(means, 97.5))


def poisson_cdf(k: int, lam: float) -> float:
    if k < 0:
        return 0.0
    term = math.exp(-lam)
    total = term
    for i in range(1, k + 1):
        term *= lam / i
        total += term
    return min(1.0, total)


def p_over(lam: float, line: float) -> float:
    return 1.0 - poisson_cdf(math.floor(line), lam)


# ---------------------------------------------------------------- A + B (ledger)

def section_a(bets: pd.DataFrame, rng) -> pd.DataFrame:
    print(f"A. Paper ROI — {len(bets):,} settled NHL paper props, flat ROI {100 * bets.pnl_flat.mean():+.1f}%\n")
    groups = [("ALL", "all", "all", bets)]
    groups += [(m, s, str(st), g) for (m, s, st), g in bets.groupby(["market", "side", "strength"])]
    groups += [(m, "any", "any", g) for m, g in bets.groupby("market")]
    groups += [("ALL", s, "any", g) for s, g in bets.groupby("side")]
    rows = []
    for market, side, strength, g in groups:
        if len(g) < MIN_BETS:
            continue
        lo, hi = boot_ci(g.pnl_flat.to_numpy(), rng)
        rows.append({"market": market, "side": side, "strength": strength, "bets": len(g),
                     "wins": int((g.result == "W").sum()), "hit_rate": float((g.result == "W").mean()),
                     "roi_mean": float(g.pnl_flat.mean()), "roi_ci_lo": lo, "roi_ci_hi": hi, "verdict": verdict(lo, hi)})
    out = pd.DataFrame(rows).sort_values("roi_ci_lo", ascending=False)
    print(f"  {'market':<26}{'side':<6}{'str':>4}{'bets':>6}{'wins':>6}{'hit%':>7}{'ROI':>8}{'CI lo':>8}{'CI hi':>8}  verdict")
    for r in out.itertuples():
        print(f"  {r.market:<26}{r.side:<6}{r.strength:>4}{r.bets:>6}{r.wins:>6}{100 * r.hit_rate:>6.1f}%{100 * r.roi_mean:>+7.1f}%"
              f"{100 * r.roi_ci_lo:>+7.1f}%{100 * r.roi_ci_hi:>+7.1f}%  {r.verdict}")
    return out


def slice_table(df: pd.DataFrame, col: str, order: list[str], title: str) -> list[dict]:
    rows = []
    print(f"\n{title}")
    print(f"  {'bucket':<26}{'n':>5}{'W':>5}{'hit%':>8}{'ROI':>8}   95% CI on hit")
    for key in [k for k in order if k in set(df[col])]:
        g = df[df[col] == key]
        if len(g) < MIN_N:
            continue
        w = int((g.result == "W").sum())
        p, lo, hi = wilson(w, len(g) - int((g.result == "P").sum()))
        rows.append({"dimension": col, "bucket": key, "n": len(g), "wins": w, "hit_rate": p, "hit_lo": lo, "hit_hi": hi,
                     "roi_flat": float(g.pnl_flat.mean())})
        print(f"  {key:<26}{len(g):>5}{w:>5}{100 * p:>7.1f}%{100 * g.pnl_flat.mean():>+7.1f}%   [{100 * lo:.0f}, {100 * hi:.0f}]")
    return rows


def section_b(bets: pd.DataFrame) -> pd.DataFrame:
    b = bets[bets.strength >= 1].copy()
    b["edge_band"] = pd.cut(b.edge, EDGE_BINS, labels=EDGE_LABELS, right=False).astype(str)
    rows = slice_table(b, "edge_band", EDGE_LABELS, "B1. by edge band (does a bigger projection-vs-line gap win more?)")
    rows += slice_table(b, "market", sorted(b.market.unique()), "B2. by market")
    rows += slice_table(b, "side", ["over", "under"], "B3. by side")
    return pd.DataFrame(rows)


# ---------------------------------------------------------------- C (calibration)

def section_c(logs: pd.DataFrame) -> pd.DataFrame:
    rows = []
    for stat, line in LINES.items():
        if stat == "saves":
            d = logs[(logs.started == 1) & logs.saves.notna()]
        else:
            d = logs[logs.toi.notna() & logs[stat].notna()]
        if d.empty:
            continue
        lg = float(d[stat].mean())
        pn = p_over(lg, line)
        bins: dict[int, list] = {}
        ll_m = ll_n = 0.0
        n = 0
        for _, grp in d.groupby("player_id", sort=False):
            vs = grp[stat].to_numpy(dtype=float)
            for i in range(MIN_PRIOR, len(vs)):
                hist = vs[:i]
                w = len(hist) / (len(hist) + SHRINK_GAMES)
                lam = w * hist.mean() + (1 - w) * lg
                rv = hist[-RECENT_GAMES:]
                if len(rv) >= 5:
                    lam = (1 - RECENT_WEIGHT) * lam + RECENT_WEIGHT * rv.mean()
                po = p_over(lam, line)
                y = 1.0 if vs[i] > line else 0.0
                eps = 1e-6
                ll_m += -(y * math.log(max(po, eps)) + (1 - y) * math.log(max(1 - po, eps)))
                ll_n += -(y * math.log(max(pn, eps)) + (1 - y) * math.log(max(1 - pn, eps)))
                b = min(9, int(po * 10))
                acc = bins.setdefault(b, [0, 0.0, 0.0])
                acc[0] += 1
                acc[1] += po
                acc[2] += y
                n += 1
        if n == 0:
            continue
        print(f"\nC. {stat.upper()} over {line:g} — {n:,} player-games ({PRIOR_SEASON}), walk-forward, no prior season")
        print(f"  log-loss: model {ll_m / n:.4f} · naive league-average {ll_n / n:.4f} → "
              f"{'model better' if ll_m < ll_n else 'NAIVE better'} by {abs(ll_m - ll_n) / n:.4f}")
        print(f"  {'P(over) bin':<12}{'n':>7}{'pred':>8}{'obs':>8}{'Δpp':>7}")
        for b in sorted(bins):
            c, sp, sy = bins[b]
            print(f"  {b / 10:.1f}-{(b + 1) / 10:.1f}     {c:>7}{100 * sp / c:>7.1f}%{100 * sy / c:>7.1f}%{100 * (sy - sp) / c:>+6.1f}")
            rows.append({"stat": stat, "line": line, "bin": f"{b / 10:.1f}-{(b + 1) / 10:.1f}", "n": c, "pred": sp / c, "obs": sy / c})
        rows.append({"stat": stat, "line": line, "bin": "logloss", "n": n, "pred": ll_m / n, "obs": ll_n / n})
    return pd.DataFrame(rows)


def main() -> None:
    OUT_DIR.mkdir(exist_ok=True)
    rng = np.random.default_rng(SEED)
    bets = load_nhl_bets()
    if bets.empty:
        print("no settled NHL paper props yet (season opens 2026-10-07) — skipping A and B")
    else:
        section_a(bets, rng).to_csv(OUT_DIR / "nhl_roi.csv", index=False)
        section_b(bets).to_csv(OUT_DIR / "nhl_slices.csv", index=False)
    logs = load_game_logs(PRIOR_SEASON)
    if logs.empty:
        print("no game logs in nhl.db — run nhl_edge.py --build")
        return
    section_c(logs).to_csv(OUT_DIR / "nhl_calibration.csv", index=False)
    print(f"\nwrote {OUT_DIR / 'nhl_*.csv'}")


if __name__ == "__main__":
    main()

File: analysis/06_nhl/nhl_loop.R
# The NHL prop analysis loop: is the projection any good, and (once the ledger fills) does
# betting it against posted lines make money?
#
#   A. Paper ROI by market x side x strength, flat $1, bootstrap 95% CI (empty pre-season).
#   B. Slices: edge band, market, side (Wilson CI on hit, flat ROI).
#   C. Walk-forward projection calibration on the stored game logs (same recipe as
#      nhl_edge.calibrate): log-loss vs naive league-average + reliability bins.
#
# Output: analysis/_out/nhl_*.csv. Mirrors nhl_loop.py — keep in lockstep.

suppressPackageStartupMessages({
  library(dplyr)
  library(boot)
})

.this_dir <- tryCatch(dirname(sys.frame(1)$ofile), error = function(e) "analysis/06_nhl")
source(file.path(.this_dir, "..", "_shared", "load_data.R"))     # wilson()
source(file.path(.this_dir, "..", "_shared", "load_nhl.R"))

OUT_DIR   <- normalizePath(file.path(.this_dir, "..", "_out"), mustWork = FALSE)
MIN_BETS  <- 10
MIN_N     <- 5
BOOT_REPS <- 5000
SEED      <- 20261007

PRIOR_SEASON  <- 20252026
SHRINK_GAMES  <- 20
RECENT_GAMES  <- 10
RECENT_WEIGHT <- 0.35
MIN_PRIOR     <- 10
LINES <- c(shots = 2.5, points = 0.5, saves = 27.5)
EDGE_BINS   <- c(0, 8, 15, 30, 50, 100000)
EDGE_LABELS <- c("0-8", "8-15", "15-30", "30-50", "50+")

dir.create(OUT_DIR, showWarnings = FALSE, recursive = TRUE)
set.seed(SEED)

verdict <- function(lo, hi) if (lo > 0) "PROFITABLE (95% CI > 0)" else if (hi < 0) "losing (95% CI < 0)" else "inconclusive"
boot_ci <- function(x) {
  b <- boot(x, function(d, i) mean(d[i]), R = BOOT_REPS)
  ci <- boot.ci(b, type = "perc")$percent
  c(lo = ci[4], hi = ci[5])
}
p_over <- function(lam, line) 1 - ppois(floor(line), lam)

# ---------------------------------------------------------------- A + B
section_a <- function(bets) {
  cat(sprintf("A. Paper ROI — %s settled NHL paper props, flat ROI %+.1f%%\n\n",
              format(nrow(bets), big.mark = ","), 100 * mean(bets$pnl_flat)))
  groups <- list(list(market = "ALL", side = "all", strength = "all", g = bets))
  for (key in split(bets, list(bets$market, bets$side, bets$strength), drop = TRUE))
    groups[[length(groups) + 1]] <- list(market = key$market[1], side = key$side[1], strength = as.character(key$strength[1]), g = key)
  for (key in split(bets, bets$market))
    groups[[length(groups) + 1]] <- list(market = key$market[1], side = "any", strength = "any", g = key)
  for (key in split(bets, bets$side))
    groups[[length(groups) + 1]] <- list(market = "ALL", side = key$side[1], strength = "any", g = key)
  rows <- lapply(groups, function(x) {
    g <- x$g
    if (nrow(g) < MIN_BETS) return(NULL)
    ci <- boot_ci(g$pnl_flat)
    data.frame(market = x$market, side = x$side, strength = x$strength, bets = nrow(g), wins = sum(g$result == "W"),
               hit_rate = mean(g$result == "W"), roi_mean = mean(g$pnl_flat), roi_ci_lo = ci[["lo"]],
               roi_ci_hi = ci[["hi"]], verdict = verdict(ci[["lo"]], ci[["hi"]]), stringsAsFactors = FALSE)
  })
  out <- bind_rows(rows) %>% arrange(desc(roi_ci_lo))
  cat(sprintf("  %-26s%-6s%4s%6s%6s%7s%8s%8s%8s  verdict\n", "market", "side", "str", "bets", "wins", "hit%", "ROI", "CI lo", "CI hi"))
  for (i in seq_len(nrow(out))) {
    r <- out[i, ]
    cat(sprintf("  %-26s%-6s%4s%6d%6d%6.1f%%%+7.1f%%%+7.1f%%%+7.1f%%  %s\n", r$market, r$side, r$strength, r$bets, r$wins,
                100 * r$hit_rate, 100 * r$roi_mean, 100 * r$roi_ci_lo, 100 * r$roi_ci_hi, r$verdict))
  }
  out
}

slice_table <- function(df, col, order, title) {
  cat(sprintf("\n%s\n", title))
  cat(sprintf("  %-26s%5s%5s%8s%8s   95%% CI on hit\n", "bucket", "n", "W", "hit%", "ROI"))
  rows <- list()
  for (key in order[order %in% unique(df[[col]])]) {
    g <- df[df[[col]] == key, ]
    if (nrow(g) < MIN_N) next
    w <- sum(g$result == "W"); wi <- wilson(w, nrow(g) - sum(g$result == "P"))
    rows[[length(rows) + 1]] <- data.frame(dimension = col, bucket = key, n = nrow(g), wins = w, hit_rate = wi[1],
                                           hit_lo = wi[2], hit_hi = wi[3], roi_flat = mean(g$pnl_flat), stringsAsFactors = FALSE)
    cat(sprintf("  %-26s%5d%5d%7.1f%%%+7.1f%%   [%.0f, %.0f]\n", key, nrow(g), w, 100 * wi[1], 100 * mean(g$pnl_flat),
                100 * wi[2], 100 * wi[3]))
  }
  rows
}

section_b <- function(bets) {
  b <- bets[bets$strength >= 1, ]
  b$edge_band <- as.character(cut(b$edge, EDGE_BINS, labels = EDGE_LABELS, right = FALSE))
  rows <- slice_table(b, "edge_band", EDGE_LABELS, "B1. by edge band (does a bigger projection-vs-line gap win more?)")
  rows <- c(rows, slice_table(b, "market", sort(unique(b$market)), "B2. by market"))
  rows <- c(rows, slice_table(b, "side", c("over", "under"), "B3. by side"))
  bind_rows(rows)
}

# ---------------------------------------------------------------- C
section_c <- function(logs) {
  rows <- list()
  for (stat in names(LINES)) {
    line <- LINES[[stat]]
    d <- if (stat == "saves") logs[!is.na(logs$started) & logs$started == 1 & !is.na(logs$saves), ] else
      logs[!is.na(logs$toi) & !is.na(logs[[stat]]), ]
    if (nrow(d) == 0) next
    lg <- mean(d[[stat]]); pn <- p_over(lg, line)
    bins <- list(); ll_m <- 0; ll_n <- 0; n <- 0
    eps <- 1e-6
    for (vs in split(d[[stat]], d$player_id)) {           # split() keeps within-player row order
      vs <- as.numeric(vs)
      if (length(vs) <= MIN_PRIOR) next
      for (i in (MIN_PRIOR + 1):length(vs)) {
        hist <- vs[1:(i - 1)]
        w <- length(hist) / (length(hist) + SHRINK_GAMES)
        lam <- w * mean(hist) + (1 - w) * lg
        rv <- tail(hist, RECENT_GAMES)
        if (length(rv) >= 5) lam <- (1 - RECENT_WEIGHT) * lam + RECENT_WEIGHT * mean(rv)
        po <- p_over(lam, line)
        y <- if (vs[i] > line) 1 else 0
        ll_m <- ll_m - (y * log(max(po, eps)) + (1 - y) * log(max(1 - po, eps)))
        ll_n <- ll_n - (y * log(max(pn, eps)) + (1 - y) * log(max(1 - pn, eps)))
        b <- as.character(min(9, floor(po * 10)))
        if (is.null(bins[[b]])) bins[[b]] <- c(0, 0, 0)
        bins[[b]] <- bins[[b]] + c(1, po, y)
        n <- n + 1
      }
    }
    if (n == 0) next
    cat(sprintf("\nC. %s over %g — %s player-games (%d), walk-forward, no prior season\n", toupper(stat), line,
                format(n, big.mark = ","), PRIOR_SEASON))
    cat(sprintf("  log-loss: model %.4f · naive league-average %.4f → %s by %.4f\n", ll_m / n, ll_n / n,
                if (ll_m < ll_n) "model better" else "NAIVE better", abs(ll_m - ll_n) / n))
    cat(sprintf("  %-12s%7s%8s%8s%7s\n", "P(over) bin", "n", "pred", "obs", "Δpp"))
    for (b in as.character(sort(as.integer(names(bins))))) {
      v <- bins[[b]]; bi <- as.integer(b)
      cat(sprintf("  %.1f-%.1f     %7d%7.1f%%%7.1f%%%+6.1f\n", bi / 10, (bi + 1) / 10, v[1], 100 * v[2] / v[1],
                  100 * v[3] / v[1], 100 * (v[3] - v[2]) / v[1]))
      rows[[length(rows) + 1]] <- data.frame(stat = stat, line = line, bin = sprintf("%.1f-%.1f", bi / 10, (bi + 1) / 10),
                                             n = v[1], pred = v[2] / v[1], obs = v[3] / v[1], stringsAsFactors = FALSE)
    }
    rows[[length(rows) + 1]] <- data.frame(stat = stat, line = line, bin = "logloss", n = n, pred = ll_m / n, obs = ll_n / n,
                                           stringsAsFactors = FALSE)
  }
  bind_rows(rows)
}

bets <- load_nhl_bets()
if (nrow(bets) == 0) {
  cat("no settled NHL paper props yet (season opens 2026-10-07) — skipping A and B\n")
} else {
  write.csv(section_a(bets), file.path(OUT_DIR, "nhl_roi.csv"), row.names = FALSE)
  write.csv(section_b(bets), file.path(OUT_DIR, "nhl_slices.csv"), row.names = FALSE)
}
logs <- load_game_logs(PRIOR_SEASON)
if (nrow(logs) == 0) {
  cat("no game logs in nhl.db — run nhl_edge.py --build\n")
  quit(status = 0)
}
write.csv(section_c(logs), file.path(OUT_DIR, "nhl_calibration.csv"), row.names = FALSE)
cat(sprintf("\nwrote %s\n", file.path(OUT_DIR, "nhl_*.csv")))
